app/src/main/java/org/jellyfin/mobile/fragment/WebViewFragment.kt

```diff
@@ -0,0 +1,198 @@
+                    path.endsWith(Constants.SESSION_CAPABILITIES_PATH) -> {
+                        runOnUiThread {
+                            webView.evaluateJavascript("window.localStorage.getItem('jellyfin_credentials')") { result ->
+                                try {
+                                    val credentials = JSONObject(result.unescapeJson())
+                                    val server = credentials.getJSONArray("Servers").getJSONObject(0)
+                                    val address = server.getString("ManualAddress")
+                                    val user = server.getString("UserId")
+                                    val token = server.getString("AccessToken")
+                                    apiClient.ChangeServerLocation(address)
+                                    apiClient.SetAuthenticationInfo(token, user)
+                                    initLocale()
+                                } catch (e: JSONException) {
+                                    Timber.e(e, "Failed to extract apiclient credentials")
+                                }
+                            }
+                        }
+                        null
+                    }
+                    else -> null
+                }
+            }
+
+            override fun onReceivedHttpError(view: WebView, request: WebResourceRequest, errorResponse: WebResourceResponse) {
+                val errorMessage = errorResponse.data?.run { bufferedReader().use(Reader::readText) }
+                Timber.e("Received WebView HTTP %d error: %s", errorResponse.statusCode, errorMessage)
+                if (request.url.path?.endsWith(Constants.INDEX_PATH) != false)
+                    runOnUiThread { onErrorReceived() }
+            }
+
+            override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceErrorCompat) {
+                val description = if (WebViewFeature.isFeatureSupported(WebViewFeature.WEB_RESOURCE_ERROR_GET_DESCRIPTION)) error.description else null
+                Timber.e("Received WebView error at %s: %s", request.url.toString(), description)
+            }
+        }
+        webChromeClient = WebChromeClient()
+        settings.apply {
+            javaScriptEnabled = true
+            domStorageEnabled = true
+        }
+        addJavascriptInterface(NativeInterface(this@WebViewFragment), "NativeInterface")
+        addJavascriptInterface(NativePlayer(context), "NativePlayer")
+        addJavascriptInterface(externalPlayer, "ExternalPlayer")
+
+        loadUrl(requireNotNull(appPreferences.instanceUrl) { "Server url has not been set!" })
+    }
+
+    fun onConnectedToWebapp() {
+        connected = true
+        (activity as? MainActivity)?.requestNoBatteryOptimizations()
+    }
+
+    fun onSelectServer() {
+        activity?.replaceFragment<ConnectFragment>()
+    }
+
+    fun onErrorReceived() {
+        connected = false
+        //showConnectionError() // TODO
```

Comment:
Is this an older TODO or a new one?

Reply:
Oops no, that's a new one. Kinda missed that, thanks for pointing it out :+1: